core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -328,19 +354,16 @@ private[spark] object TaskMetrics extends Logging {
    */
   def fromAccumulators(accums: Seq[AccumulatorV2[_, _]]): TaskMetrics = {
     val tm = new TaskMetrics
-    val externalAccums = new java.util.ArrayList[AccumulatorV2[Any, Any]]()
     for (acc <- accums) {
       val name = acc.name
-      val tmpAcc = acc.asInstanceOf[AccumulatorV2[Any, Any]]
       if (name.isDefined && tm.nameToAccums.contains(name.get)) {
         val tmAcc = tm.nameToAccums(name.get).asInstanceOf[AccumulatorV2[Any, Any]]
         tmAcc.metadata = acc.metadata
-        tmAcc.merge(tmpAcc)
+        tmAcc.merge(acc.asInstanceOf[AccumulatorV2[Any, Any]])
       } else {
-        externalAccums.add(tmpAcc)
+        tm._externalAccums += acc
```

Comment:
I didn't use the write lock by design as `tm` is a local variable so theoretically there won't be thread-safe issue for the write operation on `tm. _externalAccums`. Using write lock to be defensive is also fine given it's fast since there is no lock competition with other threads.

Reply:
Yes, `tm` is a local variable, and it is unlikely that there will be a reference to `tm` from outside before this function returns, so I also believe that there should be no thread safety issues here

Reply:
It has to do with some nits with Java memory model guarantees - the updates to the list need not be visible when queried from the read/write lock since it did not go through the same barrier when updates were made.

Reply:
> If we believe that the updates to the list are guaranteed to satisfy the happens-before, wrt any other subsequent read/write (within the lock), then we don't need to make changes - my quick review indicated this was not guaranteed, but perhaps I missed something ?

There is only the access to `nameToAccums` of `tm` other than the updates to `_externalAccums`. And the access doesn't require the read/write lock. So I think it satisfies the happens-before relationship?

Reply:
This local variable is well-scoped and I don't think it has multi-thread issues.

Reply:
@LuciferYang thanks, the publish to `LinkedBlockingQueue` is what provides the necessary happens-before relationship and so ensures visibility !